actix-web-location/src/providers.rs

```diff
@@ -0,0 +1,244 @@
+use actix_web::HttpRequest;
+use async_trait::async_trait;
+#[cfg(feature = "maxmind")]
+pub use maxmind::MaxMindProvider;
+
+/// An object that can be queried to convert [`HttpRequest`] into locations.
+///
+/// Use [`macro@async_trait`] when implementing.
+#[async_trait(?Send)]
+pub trait Provider: Send + Sync {
+    /// Provide a name of the provider for use in diagnostics.
+    fn name(&self) -> &str;
+
+    /// Derive a location from a request's metadata.
+    async fn get_location(&self, request: &HttpRequest) -> Result<Option<Location>, Error>;
+
+    /// Can this provider produce locations with country information?
+    fn expect_country(&self) -> bool {
+        true
+    }
+
+    /// Can this provider produce locations with region information?
+    fn expect_region(&self) -> bool {
+        true
+    }
+
+    /// Can this provider produce locations with city information?
+    fn expect_city(&self) -> bool {
+        true
+    }
+}
+
+/// A "dummy" provider that returns None for all fields.
+pub struct FallbackProvider {
+    fallback: Location,
+}
+
+impl FallbackProvider {
+    /// Create a fallback provider.
+    ///
+    /// The passed location builder will be modified to include the provider name.
+    pub fn new(fallback_builder: LocationBuilder) -> Self {
+        Self {
+            fallback: fallback_builder
+                .provider("fallback".to_string())
+                .finish()
+                .expect("Location construction bug"),
+        }
+    }
+}
+
+#[async_trait(?Send)]
+impl Provider for FallbackProvider {
+    fn name(&self) -> &str {
+        "fallback"
+    }
+
+    async fn get_location(&self, _request: &HttpRequest) -> Result<Option<Location>, Error> {
+        Ok(Some(self.fallback.clone()))
+    }
+}
+
+#[cfg(feature = "maxmind")]
+mod maxmind {
+    use crate::domain::LocationBuilder;
+
+    use super::{Error, Location, Provider};
+    use actix_web::{http::HeaderName, HttpRequest};
+    use anyhow::anyhow;
+    use async_trait::async_trait;
+    use lazy_static::lazy_static;
+    use maxminddb::geoip2::City;
+    use std::{net::AddrParseError, path::Path, sync::Arc};
+
+    lazy_static! {
+        static ref X_FORWARDED_FOR: HeaderName = HeaderName::from_static("x-forwarded-for");
+    }
+
+    /// A provider that uses a MaxMind GeoIP database to derive location from a the IP a request was sent from.
+    #[derive(Clone)]
+    pub struct MaxMindProvider {
+        mmdb: Arc<maxminddb::Reader<Vec<u8>>>,
+    }
+
+    impl MaxMindProvider {
+        /// Read a file from the given path into memory, and use it to construct a location provider.
+        pub fn from_path(path: &Path) -> Result<Self, Error> {
+            Ok(Self {
+                mmdb: maxminddb::Reader::open_readfile(path)
+                    .map_err(|e| Error::Setup(anyhow!("{}", e)))
+                    .map(Arc::new)?,
+            })
+        }
+    }
+
+    #[cfg(feature = "maxmind")]
```

Comment:
Isn't this already specified by the mod wrapper `[cfg]`?

Reply:
Ah, yes it is. I think I moved things around one too many times.